### ISSUE TITLE
Automatic adding of {% generation %} tags to chat_template for training with assistant_only_loss=True

### ISSUE THREAD

**Issue by Neelectric** [huggingface/trl#4879]
### Feature request

🤗Tokenizers added support for `{% generation %} {% endgeneration %}` tags, such that when using `tokenizer.apply_chat_template(..., return_assistant_tokens_mask=True)`, all non-assistant tokens are masked. This is necessary for `SFTConfig(..., assistant_only_loss=True)` to work as expected. The problem is that many existing chat templates do not yet have `{% generation %} {% endgeneration %}` tags, requiring the users to edit the chat templates correctly themselves. This feature request suggests implementing functionality such that if `assistant_only_loss=True` is passed, and the chat_template does not yet have  `{% generation %} {% endgeneration %}` tags, the model type is automatically recognised and amended appropriately. This follows suggestions by 🤗 users 'waleko' (https://huggingface.co/Qwen/Qwen3-8B/discussions/14) and 'HarryMayne' (https://github.com/HarryMayne/qwen_3_chat_templates, code below is copied and slightly adapted from HarryMayne.):

```
from tokenizers import AutoTokenizer
from trl.chat_template_utils import add_generation_tags

# Load tokenizer
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-8B")

# Automatically add generation tags
tokenizer = add_generation_tags(tokenizer)

# Apply template with assistant token masking
conversation = [
    {"role": "user", "content": "Hello assistant"},
    {"role": "assistant", "content": "Hello user"},
    {"role": "user", "content": "How are you?"},
    {"role": "assistant", "content": "I'm good"},
]

tokenized_output = tokenizer.apply_chat_template(
    conversation,
    return_assistant_tokens_mask=True,
    return_dict=True,
)

# BEFORE
# {'input_ids': [151644, 872, 198, 9707, 17847, 151645, 198, 151644, 77091, 198, 9707, 1196, 151645, 198, 151644, 872, 198, 4340, 525, 498, 30, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 40, 2776, 1661, 151645, 198], 
#  'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
#  'assistant_masks': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
# }

# AFTER
# {'input_ids': [151644, 872, 198, 9707, 17847, 151645, 198, 151644, 77091, 198, 9707, 1196, 151645, 198, 151644, 872, 198, 4340, 525, 498, 30, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 40, 2776, 1661, 151645, 198], 
#  'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
#  'assistant_masks': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1]
# }
```

### Motivation

This would allow users to train with `assistant_only_loss=True` with a simple toggle. I've found implementing the generation tags to be very fiddly and frustrating - considering most users use SFTTrainer to fine-tune a small collection of models (Llama 3.1, Qwen3 etc) it might make sense to make a utility function that makes peoples lives easier.

### Your contribution

I'm starting to work on this currently but will require support as I'm not perfectly versed in jinja tokenizer chat templates.

**Comment by qgallouedec** [comment]
Agree!
We can follow the same approach as in 

https://github.com/huggingface/trl/blob/96021dee9e38ab005a774c2cbfcc33c91e506388/trl/chat_template_utils.py#L238-L275

I'd be happy to receive a PR

**Comment by 3rdAT** [comment]
Are you also adding for Llama models? @Neelectric ?

**Comment by qgallouedec** [comment]
Hey @Neelectric, sorry for the slow reply, we're catching up on the backlog.

Good news: this is now implemented, and even more automatic than proposed. You no longer need to call any utility yourself; just set `SFTConfig(assistant_only_loss=True)` and the `SFTTrainer` detects the model's template and swaps in a `{% generation %}`-tagged training template automatically when the original lacks the markers, see https://github.com/huggingface/trl/blob/15945f9fc6d4573449fcb73dfdaef60cc985ca8a/trl/trainer/sft_trainer.py#L1239-L1244 and https://github.com/huggingface/trl/blob/15945f9fc6d4573449fcb73dfdaef60cc985ca8a/trl/chat_template_utils.py#L705-L775
Both Qwen3 and Llama 3 are covered (cc @3rdAT), along with Gemma, Phi-3, GPT-OSS, Nemotron and more. Closing as fixed.

**Comment by Neelectric** [comment]
Wow, that's fantastic! Thank you @qgallouedec , I'd kind of forgotten about this PR due to some paper deadlines 😅 great to hear that this basically works automatically now, well done folks!
